cobalt/media/base/sbplayer_pipeline.cc

```diff
@@ -1135,6 +1182,59 @@ void SbPlayerPipeline::OnNeedData(DemuxerStream::Type type,
                               type, max_buffers));
 }
 
+int SbPlayerPipeline::GetPredefinedMaxBuffers(AudioCodec codec,
+                                              TimeDelta time_allow_to_write,
+                                              bool is_preroll) {
+  // Return predefined maximum sample writes to speed up the initial sample
+  // writes, including guard number of sample writes for audio preroll.
+  int predefined_max_buffers = static_cast<int>(
+      std::ceil(time_allow_to_write.InSecondsF() *
+                audio_stream_->audio_decoder_config().samples_per_second() /
+                GetPredefinedAudioFrames(codec)));
+  if (is_preroll) {
+    predefined_max_buffers += kPrerollGuardAudioBuffer;
+  }
+  DCHECK_GT(predefined_max_buffers, 0);
+  return predefined_max_buffers;
+}
+
+int SbPlayerPipeline::GetEstimatedMaxBuffers(TimeDelta write_duration,
+                                             TimeDelta time_ahead_of_playback,
+                                             bool is_preroll) {
+  DCHECK_GE(time_ahead_of_playback.InNanoseconds(), 0);
+
+  int estimated_max_buffers = 1;
+  if (!allow_batched_sample_write_ ||
+      write_duration <= time_ahead_of_playback) {
+    return estimated_max_buffers;
+  }
+
+  TimeDelta time_allow_to_write = write_duration - time_ahead_of_playback;
+  DCHECK_GT(time_allow_to_write.InNanoseconds(), 0);
+  switch (audio_stream_->audio_decoder_config().codec()) {
+    case AudioCodec::kOpus:
+    case AudioCodec::kAAC:
+    case AudioCodec::kAC3:
+    case AudioCodec::kEAC3:
+      estimated_max_buffers =
```

Comment:
Consider only using this logic when last_audio_sample_interval_ is 0.

Reply:
Done.